File: EMS/Network/EMSSpeakersParser.m
//
//  EMSSpeakersParser.m
//

#import "EMS-Swift.h"

#import "EMSSpeakersParser.h"

#import "CJCollection.h"
#import "CJItem.h"
#import "CJLink.h"

static const DDLogLevel ddLogLevel = DDLogLevelDebug;

@implementation EMSSpeakersParser

- (NSArray *)processData:(NSData *)data forHref:(NSURL *)href error:(NSError **)error {
    NSError *parseError = nil;

    CJCollection *collection = [CJCollection collectionForNSData:data error:&parseError];

    if (!collection) {
        DDLogError(@"Failed to retrieve speakers %@ - %@ - %@", href, parseError, [parseError userInfo]);

        *error = parseError;

        return [NSArray array];
    }

    NSMutableArray *temp = [[NSMutableArray alloc] init];

    [collection.items enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
        CJItem *item = (CJItem *) obj;

        EMSSpeaker *speaker = [[EMSSpeaker alloc] init];

        speaker.href = item.href;

        [item.data enumerateObjectsUsingBlock:^(id dataObj, NSUInteger dataIdx, BOOL *dataStop) {
            NSDictionary *dict = (NSDictionary *) dataObj;

            NSString *field = dict[@"name"];
            NSString *value = dict[@"value"];

            if ([@"name" isEqualToString:field]) {
                speaker.name = value;
            }
            if ([@"bio" isEqualToString:field]) {
                speaker.bio = value;
            }
        }];

        [item.links enumerateObjectsUsingBlock:^(id linksObj, NSUInteger linksIdx, BOOL *linksStop) {
            CJLink *link = (CJLink *) linksObj;

            if ([@"thumbnail" isEqualToString:link.rel]) {
                speaker.thumbnailUrl = link.href;
            }
        }];

        speaker.lastUpdated = [NSDate date];

        [temp addObject:speaker];
    }];

    return [NSArray arrayWithArray:temp];
}

- (void)parseData:(NSData *)data forHref:(NSURL *)href {
    NSError *error;

    NSArray *collection = [self processData:data forHref:href error:&error];

    [self.delegate finishedSpeakers:collection forHref:href error:error];
}

@end

File: EMS/Network/Parsers/EMSSpeakersParser.swift
import Foundation
import CocoaLumberjack

@objc protocol EMSSpeakersParserDelegate {
    optional func finishedSpeakers(speakers: [EMSSpeaker], forHref href:NSURL, error:NSError?)
}

@objc class EMSSpeakersParser : NSObject {
    var delegate : EMSSpeakersParserDelegate?
    
    func processData(data: NSData, forHref href:NSURL) throws -> [EMSSpeaker] {
        
        do {
            let collection = try CJCollection(forNSData:data)
            
            return collection.items.map({ item in
                let speaker = EMSSpeaker()
                
                speaker.href = item.href
                
                for datum in item.data {
                    if let field = datum["name"] as? String {
                        if let value = datum["value"] as? String {
                            switch (field) {
                            case "name":
                                speaker.name = value
                            case "bio":
                                speaker.bio = value
                            default:
                                DDLogDebug("Ignored string field \(field) for speaker")
                            }
                        }
                    }
                }
                
                for link in item.links {
                    switch (link.rel) {
                    case "thumbnail":
                        speaker.thumbnailUrl = link.href
                    default:
                        DDLogDebug("Ignored rel \(link.rel) for speaker")
                    }
                }
                
                speaker.lastUpdated = NSDate()

                return speaker
            })
        } catch let error as NSError {
            DDLogError("Failed to retrieve speakers \(href) - \(error)")
            
            throw error
        }
    }
    
    func parseData(data: NSData, forHref url: NSURL) {
        do {
            let collection = try self.processData(data, forHref: url)
            
            self.delegate?.finishedSpeakers?(collection, forHref: url, error: nil)
        } catch let error as NSError {
            self.delegate?.finishedSpeakers?([], forHref: url, error: error)
        }
    }
}
